openblas@0.3.20 %oneapi@2022.1.0: libopenblas.so not linked with libifcore: undefined reference to for_cpystr, ...

### Steps to reproduce the issue

The library generated by building `openblas %oneapi` is not linked with `libifcore` and has undefined symbols as a result.

Here I am using:
* `spack/develop` (babb0edf077dfccf8e66ab63784437b06bdc3963 from `Mon Jul 25 07:50:22 2022 -0600`)
* Docker container image: `ecpe4s/ubuntu20.04-runner-x86_64-oneapi:2022-07-01`
  * Ubuntu 20.04, X86_64
  * `oneapi@2022.1.0` (installed via offline installer to default location)

Concrete spec:  [openblas.spec.json.txt](https://github.com/spack/spack/files/9183303/openblas.spec.json.txt)

Concretization:
```
[+]  openblas@0.3.20%oneapi@2022.1.0~bignuma~consistent_fpcsr~ilp64+locking+pic+shared symbol_suffix=none threads=openmp arch=linux-ubuntu20.04-x86_64
[+]      ^perl@5.34.1%oneapi@2022.1.0+cpanm+shared+threads arch=linux-ubuntu20.04-x86_64
[+]          ^berkeley-db@18.1.40%oneapi@2022.1.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-x86_64
[+]          ^bzip2@1.0.8%oneapi@2022.1.0~debug~pic+shared arch=linux-ubuntu20.04-x86_64
[+]              ^diffutils@3.8%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]                  ^libiconv@1.16%oneapi@2022.1.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
[+]          ^gdbm@1.19%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]              ^readline@8.1.2%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]                  ^ncurses@6.2%oneapi@2022.1.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-x86_64
[+]                      ^pkgconf@1.8.0%oneapi@2022.1.0 arch=linux-ubuntu20.04-x86_64
[+]          ^zlib@1.2.12%oneapi@2022.1.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
```

Compiler definition:
```
compilers:
  - compiler:
      spec: oneapi@2022.1.0
      paths:
        cc: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/icx
        cxx: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/icpx
        f77: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/ifx
        fc: /opt/intel/oneapi/compiler/2022.1.0/linux/bin/ifx
      flags: {}
      operating_system: ubuntu20.04
      target: x86_64
      modules: [compiler]
      environment: {}
      extra_rpaths: []
```

Example:
```
$> spack install -f ./openblas.spec.json

# Note libifcore is not in the list below
$> objdump -x $(spack location -i openblas)/lib/libopenblas.so | grep -i needed
  NEEDED               libimf.so
  NEEDED               libm.so.6
  NEEDED               libpthread.so.0
  NEEDED               libsvml.so
  NEEDED               libirng.so
  NEEDED               libintlc.so.5
  NEEDED               libdl.so.2
  NEEDED               libiomp5.so
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
```

As a result, when trying to build say `superlu-dist ^openblas %oneapi` you get these undefined symbol errors:
```
==> Installing superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s
==> No binary for superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/20/20b60bd8a3d88031c9ce6511ae9700b7a8dcf12e2fd704e74b1af762b3468b8c.tar.gz
==> Applied patch /spack/var/spack/repos/builtin/packages/superlu-dist/CMAKE_INSTALL_LIBDIR.patch
==> superlu-dist: Executing phase: 'cmake'
==> superlu-dist: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

79 errors found in build log:
     846    [ 82%] Linking C executable pdtest
     847    cd /tmp/root/spack-stage/spack-stage-superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/spack-build-6tn2tyk/TEST && /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0
            /cmake-3.23.2-4522tfkc3hkawky76rjynn3nkgkz3u2x/bin/cmake -E cmake_link_script CMakeFiles/pdtest.dir/link.txt --verbose=1
     848    /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-sskl3e7jafbxqzrmufsjoaxapp6r23yq/bin/mpicc -I/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/m
            etis-5.1.0-izfcwp7i63zk4ti27qteddz5wnka4nta/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/parmetis-4.0.3-7k7kmefyaoqnhigelovllym5ixi4dxjo/include -DUSE_
            VENDOR_BLAS  -std=c99 -O2 -g -DNDEBUG  -Wl,-rpath,/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/SRC   -Wl,-rpa
            th,/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s//spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/sup
            erlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/lib CMakeFiles/pdtest.dir/pdtest.c.o CMakeFiles/pdtest.dir/dcreate_matrix.c.o CMakeFiles/pdtest.dir/pdcompute_resid.c.o -o pd
            test  -Wl,-rpath,/tmp/root/spack-stage/spack-stage-superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/spack-build-6tn2tyk/SRC:/spack/opt/spack/linux-ubuntu20.04-x86_64/oneap
            i-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib ../SRC/libsuperlu_dist.so.7.2.0 /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdg
            wsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so -lm -L/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/parmetis-4.0.3-7k7kmefyaoqnhigelovllym5ixi4dxjo/lib -lparmetis
             -L/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/metis-5.1.0-izfcwp7i63zk4ti27qteddz5wnka4nta/lib -lmetis
     849    [ 82%] Linking C executable pddrive
     850    cd /tmp/root/spack-stage/spack-stage-superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/spack-build-6tn2tyk/EXAMPLE && /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.
            1.0/cmake-3.23.2-4522tfkc3hkawky76rjynn3nkgkz3u2x/bin/cmake -E cmake_link_script CMakeFiles/pddrive.dir/link.txt --verbose=1
     851    /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/mpich-4.0.2-sskl3e7jafbxqzrmufsjoaxapp6r23yq/bin/mpicc -I/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/m
            etis-5.1.0-izfcwp7i63zk4ti27qteddz5wnka4nta/include -I/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/parmetis-4.0.3-7k7kmefyaoqnhigelovllym5ixi4dxjo/include -DUSE_
            VENDOR_BLAS  -std=c99 -O2 -g -DNDEBUG  -Wl,-rpath,/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/SRC   -Wl,-rpa
            th,/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s//spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/sup
            erlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/lib CMakeFiles/pddrive.dir/pddrive.c.o CMakeFiles/pddrive.dir/dcreate_matrix.c.o -o pddrive  -Wl,-rpath,/tmp/root/spack-stage
            /spack-stage-superlu-dist-7.2.0-6tn2tykzswttzpcyk3agkxnsukdg5o3s/spack-build-6tn2tyk/SRC:/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3
            gzijkrgq2panhwwcs2q6z/lib:::: ../SRC/libsuperlu_dist.so.7.2.0 /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/l
            ibopenblas.so -lm -L/spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/parmetis-4.0.3-7k7kmefyaoqnhigelovllym5ixi4dxjo/lib -lparmetis -L/spack/opt/spack/linux-ubuntu20
            .04-x86_64/oneapi-2022.1.0/metis-5.1.0-izfcwp7i63zk4ti27qteddz5wnka4nta/lib -lmetis
  >> 852    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so: undefined reference to `for_cpystr
            '
  >> 853    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so: undefined reference to `for_concat
            '
  >> 854    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so: undefined reference to `for_cpstr'
  >> 855    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so: undefined reference to `for_cpystr
            '
  >> 856    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/oneapi-2022.1.0/openblas-0.3.20-ygdgwsa2sz3gzijkrgq2panhwwcs2q6z/lib/libopenblas.so: undefined reference to `for_concat
            '
...
```

These symbols are defined in `libifcore.so.5`
```
$> nm -D /opt/intel/oneapi/compiler/2022.1.0/linux/compiler/lib/ia32_lin/libifcore.so.5 | grep -e for_cpstr -e for_concat -e for_cpystr
000ad1b0 T for_concat
000acfa0 T for_cpstr
000ac620 T for_cpstr_eq
000aca60 T for_cpstr_ge
000ac0f0 T for_cpstr_gt
000acd00 T for_cpstr_le
000ac380 T for_cpstr_lt
000ac840 T for_cpstr_ne
000ad490 T for_cpystr
```

Does anyone know why `libifcore` isn't being linked in when building `openblas`?

Similar issues from 2019:
* https://github.com/spack/spack/issues/8901
* https://github.com/xianyi/OpenBLAS/issues/1956

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (babb0edf077dfccf8e66ab63784437b06bdc3963)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

### Additional information

Build logs from `spack install -f ./openblas.spec.json` (attached above)
[spack-build-env.txt](https://github.com/spack/spack/files/9183358/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/9183359/spack-build-out.txt)

@rscohn2 @haampie @michaelkuhn @sethrj @siko1056 @martin-frbg @s-sajid-ali @becker33 @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/openblas/f_check-oneapi.patch
+++ b/var/spack/repos/builtin/packages/openblas/f_check-oneapi.patch
@@ -1,0 +1,31 @@
+diff -ruN spack-src/f_check spack-src-new/f_check
+--- spack-src/f_check	2022-02-20 13:35:05.000000000 -0800
++++ spack-src-new/f_check	2022-07-25 14:41:31.877584474 -0700
+@@ -34,7 +34,7 @@
+ 	      "pathf90", "pathf95",
+ 	      "pgf95", "pgf90", "pgf77", "pgfortran", "nvfortran",
+ 	      "flang", "egfortran",
+-              "ifort", "nagfor");
++              "ifort", "nagfor", "ifx");
+ 
+ OUTER:
+     foreach $lists (@lists) {
+@@ -90,6 +90,9 @@
+ 		if ($compiler =~ /flang/) {
+ 		    $vendor = FLANG;
+ 		    $openmp = "-fopenmp";
++	    } elsif ($compiler =~ /ifx/) {
++		    $vendor = INTEL;
++		    $openmp = "-fopenmp";
+ 	    } elsif ($compiler =~ /pgf/ || $compiler =~ /nvf/) {
+ 		    $vendor = PGI;
+ 		    $openmp = "-mp";
+@@ -176,7 +179,7 @@
+ 	    $openmp = "-fopenmp";
+ 	}
+ 
+-	if ($compiler =~ /ifort/) {
++	if ($compiler =~ /ifort/ || $compiler =~ /ifx/) {
+ 	    $vendor = INTEL;
+ 	    $bu       = "_";
+ 	    $openmp = "-fopenmp";

--- a/var/spack/repos/builtin/packages/openblas/package.py
+++ b/var/spack/repos/builtin/packages/openblas/package.py
@@ -69,6 +69,9 @@ class Openblas(MakefilePackage):
     provides('lapack@3.9.1:', when='@0.3.15:')
     provides('lapack@3.7.0', when='@0.2.20')
 
+    # https://github.com/spack/spack/issues/31732
+    patch('f_check-oneapi.patch', when='@0.3.20 %oneapi')
+
     # OpenBLAS >=3.0 has an official way to disable internal parallel builds
     patch('make.patch', when='@0.2.16:0.2.20')
     #  This patch is in a pull request to OpenBLAS that has not been handled

